File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixSums = [0] * len(nums)
        prefixSums[0] = nums[0]
        for i in range(1, len(nums)):
            prefixSums[i] = prefixSums[i-1] * nums[i]

        suffixSums = [0] * len(nums)
        suffixSums[len(nums) - 1] = nums[len(nums) - 1]
        for i in range(len(nums) - 2, -1, -1):
            suffixSums[i] = suffixSums[i+1] * nums[i]

        ans = []

        for i in range(len(nums)):
            prefixSum = 0
            suffixSum = 0
            left = i - 1
            right = i + 1
            if left < 0:
                prefixSum = 1
            else:
                prefixSum = prefixSums[left]
            
            if right > len(nums) - 1:
                suffixSum = 1
            else:
                suffixSum = suffixSums[right]

            ans.append(prefixSum * suffixSum)

        return ans

File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        return searchHelper(nums, 0, nums.size() - 1, target);
    }

    int searchHelper(vector<int>& nums, int left, int right, int target) {
        if (left > right) return -1;
        int middle = (left + right) / 2;

        if (nums[middle] > target) return searchHelper(nums, left, middle-1, target);
        else if (nums[middle] < target) return searchHelper(nums, middle + 1, right, target);
        else return middle;
    }
};
